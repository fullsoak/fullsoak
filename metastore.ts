const store = new Map();

const FULLSOAK_APP_COMPONENTS_DIR = "FULLSOAK_APP_COMPONENTS_DIR";

const setGlobalComponentsDir = (dir: string) =>
  store.set(FULLSOAK_APP_COMPONENTS_DIR, dir);
const getGlobalComponentsDir = store.get(FULLSOAK_APP_COMPONENTS_DIR);

export { getGlobalComponentsDir, setGlobalComponentsDir };
