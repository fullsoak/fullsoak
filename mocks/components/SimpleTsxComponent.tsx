export const SimpleTsxComponent = () => <div>SimpleTsxComponent</div>;
