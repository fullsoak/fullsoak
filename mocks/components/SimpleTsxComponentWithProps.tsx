import type { FunctionComponent } from "preact";

type TheProps = { name: string };
export const SimpleTsxComponentWithProps: FunctionComponent<TheProps> = (
  { name },
) => <div>SimpleTsxComponentWithProps={name}</div>;
