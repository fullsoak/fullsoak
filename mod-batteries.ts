/** @module batteries */

export { HtmlShell } from "./src/HtmlShell.ts";
export { byoHtml } from "./src/ssr.ts";
export { byoHat, Hat } from "./src/miscellaneous/Hat.ts";
export * from "./src/minifyCss.ts";
