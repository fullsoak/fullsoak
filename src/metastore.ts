import type { ApplicationListenEvent } from "@oak/oak/application";
import { LogDebug } from "./utils.ts";

const store = new Map();

const FULLSOAK_APP_COMPONENTS_DIR = "FULLSOAK_APP_COMPONENTS_DIR";
const APP_LISTEN_OBJ = "APP_LISTEN_OBJ";
export const setGlobalComponentsDir = (dir: string) => {
  store.set(FULLSOAK_APP_COMPONENTS_DIR, dir);
  LogDebug("set global components dir", dir);
};

export const getGlobalComponentsDir = (): string =>
  store.get(FULLSOAK_APP_COMPONENTS_DIR);

export const setAppListenObj = (obj: ApplicationListenEvent) =>
  store.set(APP_LISTEN_OBJ, obj);

const getAppListenObj = (): ApplicationListenEvent => store.get(APP_LISTEN_OBJ);

/**
 * server-side only
 */
export const getOrigin = (): string => {
  const obj = getAppListenObj();
  const scheme = obj.secure ? "https" : "http";
  const localOrigin = obj.port != null
    ? `${scheme}://${obj.hostname}:${obj.port}`
    : obj.hostname;
  return localOrigin;
};
