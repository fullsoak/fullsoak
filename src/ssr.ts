import type { FunctionComponent, VNode } from "preact";
import type { CP } from "./types.ts";
import { renderToStringAsync } from "preact-render-to-string";
import { withHtmlShell } from "./HtmlShell.tsx";
import { cleanCss } from "./minifyCss.ts";
import { html } from "htm/preact";
import { getComponentCss } from "./getComponentCss.ts";

/**
 * takes a component that starts from the root `<html>` element
 * and renders it out as a string; the input component is usually
 * constructed by using the {@link HtmlShell} tsx component (included battery)
 *
 * @example
 * ```tsx
 * import { byoHtml, HtmlShell } from "@fullsoak/fullsoak";
 * byoHtml(
 *   <HtmlShell
 *     componentName="MyApp"
 *     js="some vanilla javascript"
 *     css="some raw css"
 *   />
 * );
 * ```
 */
export const byoHtml = async (component: VNode): Promise<string> => {
  globalThis.location = {} as Location;
  return "<!doctype html>" + (await renderToStringAsync(component));
};

const ssrVNode = async (
  node: VNode,
): Promise<string> => {
  const componentName = typeof node.type === "string"
    ? node.type
    : node.type.name;
  const componentCss = await getComponentCss(componentName);
  return byoHtml(withHtmlShell({
    componentName,
    componentProps: node.props,
    component: node,
    css: cleanCss(componentCss),
  }));
};

const ssrTsxComponent = async <P extends CP>(
  component: FunctionComponent<P>,
  props: P | null = null,
): Promise<string> => {
  const componentName = component.name;

  // this is the only way to render a TSX component where the
  // deepest DOM element is rendered (which is the whole point of SSR)
  const componentVNode = html`<${component} ...${props} />`;

  const componentCss = await getComponentCss(componentName);
  return byoHtml(withHtmlShell({
    componentName,
    componentProps: props,
    // while using `h` also works, it doesn't render the deepest DOM element
    // component: h<CP>(componentName, props),
    component: componentVNode,
    css: cleanCss(componentCss),
  }));
};

/**
 * takes a Preact `VNode` or a TSX component and wraps it around an `<html>` element
 * then renders everything out as a string
 * @param renderTarget - a Preact VNode or a TSX component
 * @param componentProps - the props to pass to the component (only applicable when passing a TSX component as `renderTarget`)
 * @returns - the rendered HTML as a string
 * @TODO update the README example
 */
export const ssr = <P extends CP>(
  renderTarget: FunctionComponent<P> | VNode,
  componentProps: P | null = null,
): Promise<string> => {
  // @TODO see if there's a better way to check if it's a VNode
  const isVNode = typeof renderTarget === "object" && "type" in renderTarget;
  if (isVNode) return ssrVNode(renderTarget);
  return ssrTsxComponent(renderTarget, componentProps);
};
