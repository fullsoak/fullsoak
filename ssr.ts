import type { FunctionComponent, VNode } from "preact";
import { render } from "preact-render-to-string";
import { withHtmlShell } from "./HtmlShell.tsx";
import { cleanCss } from "./minifyCss.ts";
import { html } from "htm/preact";
import { getGlobalComponentsDir } from "./metastore.ts";

/**
 * takes a component that starts from the 'root' <html> element
 * and render it out as a string
 */
export const byoHtml = (component: VNode): string =>
  "<!doctype html>" + render(component);

/**
 * takes a standard Preact component and wrap it around an <html> element
 * and render it out as a string
 */
export const ssr = async (component: FunctionComponent): Promise<string> => {
  const componentName = component.name;
  const componentVNode = html`<${component} />`;
  let componentCss;
  try {
    // @TODO use a framework smart fn that attempts to read all .css files in the `componentName` dir?
    // @TODO also consider the possibility to combine a general 'main.css' and a component-specific css
    componentCss = await Deno.readTextFile(
      `${getGlobalComponentsDir()}/${componentName}/styles.css`,
    );
  } catch (e) {
    console.warn(
      `unable to load css for component ${componentName}: ${
        (e as Error).message
      }`,
    );
  }
  return byoHtml(withHtmlShell({
    component: componentVNode,
    css: cleanCss(componentCss),
  }));
};
