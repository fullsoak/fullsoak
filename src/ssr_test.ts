import { ssr } from "./ssr.ts";
import { html } from "htm/preact";
import { SimpleTsxComponent } from "@mocks/components/SimpleTsxComponent.tsx";
import { SimpleTsxComponentWithProps } from "@mocks/components/SimpleTsxComponentWithProps.tsx";
import { StandardComponent } from "@mocks/components/StandardComponent/index.tsx";
import { assertSnapshot } from "@std/testing/snapshot";
import { setGlobalComponentsDir } from "./metastore.ts";
import { Hat, makeHat } from "../mod-batteries.ts";

// this is called in `useFullsoak`, but we call it here to make unit tests simpler
setGlobalComponentsDir("./mocks/components");

Deno.test("ssr a TSX component", async (t) => {
  const output = await ssr(SimpleTsxComponent);
  await assertSnapshot(t, output);
});

Deno.test("ssr a TSX component with props", async (t) => {
  const output = await ssr(SimpleTsxComponentWithProps, { name: "Alice" });
  await assertSnapshot(t, output);
});

Deno.test("ssr a VNode using HTM", async (t) => {
  const props = { name: "Bob" };
  const output = await ssr(
    html`<${SimpleTsxComponentWithProps} ...${props} />`,
  );
  await assertSnapshot(t, output);
});

Deno.test("ssr a StandardComponent with makeHat", async (t) => {
  const output = await ssr(StandardComponent, { name: "Charlie" }, {
    headContent: makeHat({ title: "Standard Component" }),
  });
  await assertSnapshot(t, output);
});

Deno.test("ssr a StandardComponent with Hat", async (t) => {
  const hatProps = {
    title: "SC2",
    meta: [{
      name: "og:title",
      content: "SC2",
    }],
  };
  const output = await ssr(StandardComponent, { name: "Charlie" }, {
    headContent: html`<${Hat} ...${hatProps} />`,
  });
  await assertSnapshot(t, output);
});
