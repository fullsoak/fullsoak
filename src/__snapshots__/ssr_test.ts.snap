export const snapshot = {};

snapshot[`ssr a TSX component 1`] = `
'<!doctype html><html><head><meta charset="utf-8"/><meta name="viewport" content="width=device-width,initial-scale=1.0,maximum-scale=1.0"/><script type="importmap">{
    "imports": {
      "preact": "https://esm.sh/preact@10.26.4",
      "preact/hooks": "https://esm.sh/preact@10.26.4/hooks",
      "preact/debug": "https://esm.sh/preact@10.26.4/debug",
      "preact/devtools": "https://esm.sh/preact@10.26.4/devtools",
      "react": "https://esm.sh/preact@10.26.4",
      "react/jsx-runtime": "https://esm.sh/preact@10.26.4/jsx-runtime",
      "preact/jsx-runtime": "https://esm.sh/preact@10.26.4/jsx-runtime",
      "react-dom": "https://esm.sh/preact@10.26.4/compat/",
      "react-dom/*": "https://esm.sh/preact@10.26.4/compat/*",
      "htm/preact": "https://esm.sh/htm@3.1.1/preact?external=preact",
      "preact-iso": "https://esm.sh/preact-iso@^2.9.0?external=preact",
      "preact-iso/prerender": "/fullsoak",
      "@preact/signals": "https://esm.sh/@preact/signals",
      "fullsoak/preact-iso": "https://esm.sh/preact-iso@^2.9.0?external=preact",
      "@fullsoak/fullsoak/preact-iso": "https://esm.sh/preact-iso@^2.9.0?external=preact",
      "fullsoak": "/fullsoak",
      "@fullsoak/fullsoak": "/fullsoak"
    }
  }</script><link rel="modulepreload" href="/components/SimpleTsxComponent/mount" as="script" type="text/javascript"/><link rel="modulepreload" href="/components/SimpleTsxComponent/index.tsx" as="script" type="text/javascript"/><script type="text/javascript"></script><style></style></head><body><main id="fullsoak-htmlshell-main"><div>SimpleTsxComponent</div></main><script type="text/javascript">window.preloadedProps = {}</script><script defer type="module" src="/components/SimpleTsxComponent/mount"></script></body></html>'
`;

snapshot[`ssr a TSX component with props 1`] = `
'<!doctype html><html><head><meta charset="utf-8"/><meta name="viewport" content="width=device-width,initial-scale=1.0,maximum-scale=1.0"/><script type="importmap">{
    "imports": {
      "preact": "https://esm.sh/preact@10.26.4",
      "preact/hooks": "https://esm.sh/preact@10.26.4/hooks",
      "preact/debug": "https://esm.sh/preact@10.26.4/debug",
      "preact/devtools": "https://esm.sh/preact@10.26.4/devtools",
      "react": "https://esm.sh/preact@10.26.4",
      "react/jsx-runtime": "https://esm.sh/preact@10.26.4/jsx-runtime",
      "preact/jsx-runtime": "https://esm.sh/preact@10.26.4/jsx-runtime",
      "react-dom": "https://esm.sh/preact@10.26.4/compat/",
      "react-dom/*": "https://esm.sh/preact@10.26.4/compat/*",
      "htm/preact": "https://esm.sh/htm@3.1.1/preact?external=preact",
      "preact-iso": "https://esm.sh/preact-iso@^2.9.0?external=preact",
      "preact-iso/prerender": "/fullsoak",
      "@preact/signals": "https://esm.sh/@preact/signals",
      "fullsoak/preact-iso": "https://esm.sh/preact-iso@^2.9.0?external=preact",
      "@fullsoak/fullsoak/preact-iso": "https://esm.sh/preact-iso@^2.9.0?external=preact",
      "fullsoak": "/fullsoak",
      "@fullsoak/fullsoak": "/fullsoak"
    }
  }</script><link rel="modulepreload" href="/components/SimpleTsxComponentWithProps/mount" as="script" type="text/javascript"/><link rel="modulepreload" href="/components/SimpleTsxComponentWithProps/index.tsx" as="script" type="text/javascript"/><script type="text/javascript"></script><style></style></head><body><main id="fullsoak-htmlshell-main"><div>SimpleTsxComponentWithProps=Alice</div></main><script type="text/javascript">window.preloadedProps = {"name":"Alice"}</script><script defer type="module" src="/components/SimpleTsxComponentWithProps/mount"></script></body></html>'
`;

snapshot[`ssr a VNode using HTM 1`] = `
'<!doctype html><html><head><meta charset="utf-8"/><meta name="viewport" content="width=device-width,initial-scale=1.0,maximum-scale=1.0"/><script type="importmap">{
    "imports": {
      "preact": "https://esm.sh/preact@10.26.4",
      "preact/hooks": "https://esm.sh/preact@10.26.4/hooks",
      "preact/debug": "https://esm.sh/preact@10.26.4/debug",
      "preact/devtools": "https://esm.sh/preact@10.26.4/devtools",
      "react": "https://esm.sh/preact@10.26.4",
      "react/jsx-runtime": "https://esm.sh/preact@10.26.4/jsx-runtime",
      "preact/jsx-runtime": "https://esm.sh/preact@10.26.4/jsx-runtime",
      "react-dom": "https://esm.sh/preact@10.26.4/compat/",
      "react-dom/*": "https://esm.sh/preact@10.26.4/compat/*",
      "htm/preact": "https://esm.sh/htm@3.1.1/preact?external=preact",
      "preact-iso": "https://esm.sh/preact-iso@^2.9.0?external=preact",
      "preact-iso/prerender": "/fullsoak",
      "@preact/signals": "https://esm.sh/@preact/signals",
      "fullsoak/preact-iso": "https://esm.sh/preact-iso@^2.9.0?external=preact",
      "@fullsoak/fullsoak/preact-iso": "https://esm.sh/preact-iso@^2.9.0?external=preact",
      "fullsoak": "/fullsoak",
      "@fullsoak/fullsoak": "/fullsoak"
    }
  }</script><link rel="modulepreload" href="/components/SimpleTsxComponentWithProps/mount" as="script" type="text/javascript"/><link rel="modulepreload" href="/components/SimpleTsxComponentWithProps/index.tsx" as="script" type="text/javascript"/><script type="text/javascript"></script><style></style></head><body><main id="fullsoak-htmlshell-main"><div>SimpleTsxComponentWithProps=Bob</div></main><script type="text/javascript">window.preloadedProps = {"name":"Bob"}</script><script defer type="module" src="/components/SimpleTsxComponentWithProps/mount"></script></body></html>'
`;

snapshot[`ssr a StandardComponent with makeHat 1`] = `
'<!doctype html><html><head><meta charset="utf-8"/><meta name="viewport" content="width=device-width,initial-scale=1.0,maximum-scale=1.0"/><title>Standard Component</title><script type="importmap">{
    "imports": {
      "preact": "https://esm.sh/preact@10.26.4",
      "preact/hooks": "https://esm.sh/preact@10.26.4/hooks",
      "preact/debug": "https://esm.sh/preact@10.26.4/debug",
      "preact/devtools": "https://esm.sh/preact@10.26.4/devtools",
      "react": "https://esm.sh/preact@10.26.4",
      "react/jsx-runtime": "https://esm.sh/preact@10.26.4/jsx-runtime",
      "preact/jsx-runtime": "https://esm.sh/preact@10.26.4/jsx-runtime",
      "react-dom": "https://esm.sh/preact@10.26.4/compat/",
      "react-dom/*": "https://esm.sh/preact@10.26.4/compat/*",
      "htm/preact": "https://esm.sh/htm@3.1.1/preact?external=preact",
      "preact-iso": "https://esm.sh/preact-iso@^2.9.0?external=preact",
      "preact-iso/prerender": "/fullsoak",
      "@preact/signals": "https://esm.sh/@preact/signals",
      "fullsoak/preact-iso": "https://esm.sh/preact-iso@^2.9.0?external=preact",
      "@fullsoak/fullsoak/preact-iso": "https://esm.sh/preact-iso@^2.9.0?external=preact",
      "fullsoak": "/fullsoak",
      "@fullsoak/fullsoak": "/fullsoak"
    }
  }</script><link rel="modulepreload" href="/components/StandardComponent/mount" as="script" type="text/javascript"/><link rel="modulepreload" href="/components/StandardComponent/index.tsx" as="script" type="text/javascript"/><script type="text/javascript"></script><style>@import url(/components/Shared/styles.css);.foo{background-color:#55c456}</style></head><body><main id="fullsoak-htmlshell-main"><div>StandardComponent=Charlie</div></main><script type="text/javascript">window.preloadedProps = {"name":"Charlie"}</script><script defer type="module" src="/components/StandardComponent/mount"></script></body></html>'
`;

snapshot[`ssr a StandardComponent with Hat 1`] = `
'<!doctype html><html><head><meta charset="utf-8"/><meta name="viewport" content="width=device-width,initial-scale=1.0,maximum-scale=1.0"/><title>SC2</title><meta name="og:title" content="SC2"/><script type="importmap">{
    "imports": {
      "preact": "https://esm.sh/preact@10.26.4",
      "preact/hooks": "https://esm.sh/preact@10.26.4/hooks",
      "preact/debug": "https://esm.sh/preact@10.26.4/debug",
      "preact/devtools": "https://esm.sh/preact@10.26.4/devtools",
      "react": "https://esm.sh/preact@10.26.4",
      "react/jsx-runtime": "https://esm.sh/preact@10.26.4/jsx-runtime",
      "preact/jsx-runtime": "https://esm.sh/preact@10.26.4/jsx-runtime",
      "react-dom": "https://esm.sh/preact@10.26.4/compat/",
      "react-dom/*": "https://esm.sh/preact@10.26.4/compat/*",
      "htm/preact": "https://esm.sh/htm@3.1.1/preact?external=preact",
      "preact-iso": "https://esm.sh/preact-iso@^2.9.0?external=preact",
      "preact-iso/prerender": "/fullsoak",
      "@preact/signals": "https://esm.sh/@preact/signals",
      "fullsoak/preact-iso": "https://esm.sh/preact-iso@^2.9.0?external=preact",
      "@fullsoak/fullsoak/preact-iso": "https://esm.sh/preact-iso@^2.9.0?external=preact",
      "fullsoak": "/fullsoak",
      "@fullsoak/fullsoak": "/fullsoak"
    }
  }</script><link rel="modulepreload" href="/components/StandardComponent/mount" as="script" type="text/javascript"/><link rel="modulepreload" href="/components/StandardComponent/index.tsx" as="script" type="text/javascript"/><script type="text/javascript"></script><style>@import url(/components/Shared/styles.css);.foo{background-color:#55c456}</style></head><body><main id="fullsoak-htmlshell-main"><div>StandardComponent=Charlie</div></main><script type="text/javascript">window.preloadedProps = {"name":"Charlie"}</script><script defer type="module" src="/components/StandardComponent/mount"></script></body></html>'
`;
